pkg/urbit/include/c/defs.h

```diff
@@ -157,4 +159,108 @@
 #     define c3_fopen(a, b) ({                                  \
         fopen(a, b);})
 
+      /// Reads the contents of a file into memory.
+      ///
+      /// This function correctly handles the case in which read() reads fewer
+      /// bytes than was requested by retrying until all bytes have been read or
+      /// until an error occurs, whichever comes first.
+      ///
+      /// If interrupted by a signal or if the read() call would block, this
+      /// function will retry up to a specified maximum number of attempts.
+      ///
+      /// @param[in]  fd_i        Open file descriptor to read from.
+      /// @param[out] data_v      Destination buffer.
+      /// @param[in]  data_len_i  Length of `data_v` in bytes.
+      ///
+      /// @return 1  Complete read succeeded.
+      /// @return 0  Otherwise.
+      static inline c3_t
+      c3_read(c3_i fd_i, void* const data_v, const size_t data_len_i)
+      {
+        static const size_t max_attempts_i    = 100;
+        size_t              attempts_i        = 0;
+        c3_y*               ptr_y             = data_v;
```

Comment:
`<type> *const <name>`  indicates that the pointer identified by `<name>` will not change i.e. `<name>++`, `<name> += 3`, etc are all invalid. So declaring `data_v` as `void* const` indicates to the caller that we're not performing arithmetic on `data_v`. In other words, the `const`ness is tied to the variable name, not to the data that the variable points at, and as a result, creating a new variable that points to the same address does not result in discarding the `const` qualifier.

Reply:
So you could argue that it's unnecessary to declare `data_v` as `void* const`, but it provides just a little more context to whoever is reading the function definition.

Reply:
I understand the what, but not the why. Conveying to callers that data behind the pointer can/will not be modified seems very valuable, but conveying that the pointer itself will not be changed (because it will be copied first) seems irrelevant. Callers cannot be affected either way.

I've used `const` on scalars in function signatures, as a correctness constraint for the implementation. Even that  seems like it leaks some implementation details (again, because callers can't be impacted either way). But this seems like unnecessary (or even misleading) ceremony.

Reply:
This convention adheres to a school of thought in which anything that can be marked `const` should be. The pointer doesn't need to be modified in the function, and so it should be marked `const` to convey that. One word for this is ceremony, but another is discipline. We don't lose anything by marking `data_v` as `void* const` rather than `void*`, but we gain a more explicit explanation of how `data_v` is used.

Reply:
But it can only be marked `const` because it's copied before it's modified -- by that logic, everything can be `const` (otherwise known as SSA form)! So it's still not clear to me why there is value in what is being conveyed here.

Reply:
Would it feel better if I implemented `c3_read()` using a base pointer and an offset rather than a current pointer? Because in that case, then the pointer wouldn't be modified, and this argument against labeling `data_v` as `const` would no longer be relevant.

Reply:
No, the question I've put here in various forms would still remain, unanswered. "If you followed the rule, and it brought you to 'manually convert your code to SSA-form', of what use was the rule?".

More here: https://softwareengineering.stackexchange.com/questions/204500/when-and-for-what-purposes-should-the-const-keyword-be-used-in-c-for-variables

Reply:
Pointer variables marked `<type>* const` cannot be reallocated/freed without a cast. Because `data_v` is marked as an output parameter in the interface documentation of `c3_read()`, you could argue that it's obvious that `data_v` won't be reallocated/freed by the `c3_read()` implementation, but marking `data_v` as `const` in this way explains to the caller that the code encapsulates the notion of "`data_v` will not be reallocated/freed", thereby reinforcing the guarantee made by the documentation. Of course, the `const` label *can* easily be overcome by a cast or by creating a new variable (as is done here), but it's unlikely that anyone modifying the implementation of `c3_read()` will be able to overlook both the interface documentation and the fact that `data_v` is marked `const` to reallocate/free the memory owned by `data_v`. In other words, I see absolutely no reason why the `const` label should be omitted given that it communicates valuable information to both the caller and the implementer.